flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/snapshot/StateSnapshotReconciler.java

```diff
@@ -0,0 +1,230 @@
+        var resource = ctx.getResource();
+
+        var savepointState = resource.getStatus().getState();
+        if (!TRIGGER_PENDING.equals(savepointState)) {
+            return;
+        }
+
+        if (resource.getSpec().isSavepoint()
+                && resource.getSpec().getSavepoint().getAlreadyExists()) {
+            LOG.info(
+                    "Snapshot {} is marked as completed in spec, skipping triggering savepoint.",
+                    resource.getMetadata().getName());
+
+            FlinkStateSnapshotUtils.snapshotSuccessful(
+                    resource, resource.getSpec().getSavepoint().getPath(), true);
+            return;
+        }
+
+        if (FlinkStateSnapshotUtils.abandonSnapshotIfJobNotRunning(
+                ctx.getKubernetesClient(),
+                ctx.getResource(),
+                ctx.getSecondaryResource().orElse(null),
+                eventRecorder)) {
+            return;
+        }
+
+        var jobId = ctx.getSecondaryResource().orElseThrow().getStatus().getJobStatus().getJobId();
+
+        Optional<String> triggerIdOpt;
+        try {
+            triggerIdOpt = triggerCheckpointOrSavepoint(resource.getSpec(), ctx, jobId);
+        } catch (Exception e) {
+            LOG.error("Failed to trigger snapshot for resource {}", ctx.getResource(), e);
+            throw new ReconciliationException(e);
+        }
+
+        if (triggerIdOpt.isEmpty()) {
+            LOG.warn("Failed to trigger snapshot {}", resource.getMetadata().getName());
+            return;
+        }
+
+        FlinkStateSnapshotUtils.snapshotInProgress(resource, triggerIdOpt.get());
+    }
+
+    public DeleteControl cleanup(FlinkStateSnapshotContext ctx) throws Exception {
+        var resource = ctx.getResource();
+        var state = resource.getStatus().getState();
+        var resourceName = resource.getMetadata().getName();
+        LOG.info("Cleaning up resource {}...", resourceName);
+
+        if (resource.getSpec().isCheckpoint()) {
+            return DeleteControl.defaultDelete();
+        }
+        if (!resource.getSpec().getSavepoint().getDisposeOnDelete()) {
+            return DeleteControl.defaultDelete();
+        }
+        if (resource.getSpec().getJobReference() == null
+                || resource.getSpec().getJobReference().getName() == null) {
+            return DeleteControl.defaultDelete();
+        }
+
+        switch (state) {
+            case IN_PROGRESS:
+                LOG.info(
+                        "Cannot delete resource {} yet as savepoint is still in progress...",
+                        resourceName);
+                return DeleteControl.noFinalizerRemoval()
+                        .rescheduleAfter(ctx.getOperatorConfig().getReconcileInterval().toMillis());
+            case COMPLETED:
+                var flinkDeployment = getFlinkDeployment(ctx);
+                return handleSnapshotCleanup(resource, flinkDeployment, ctx);
```

Comment:
I was trying out the periodic snapshot feature today. Does it make sense to introduce a feature to auto-cleanup the  `flinkstatesnapshots.flink.apache.org` CRs and only retain upto some max retention limit (we could still retain the savepoint dir?) IIUC `disposeOnDelete` for periodic savepoints is set to false by default (so it should be relatively safe to do so). 

If the user sets the savepoint interval `kubernetes.operator.periodic.savepoint.interval` to a low value, and there are multiple jobs on the cluster it would lead to creation of many CRs. I am not familiar with K8s internals but could this increase the load on the API server?

Reply:
Definitely! I would not feel comfortable releasing a new operator version that does not have auto-cleanup of FlinkStateSnapshot CRs. For the now-deprecated way of taking snapshots this could be configured using `kubernetes.operator.savepoint.history.max.age` and `kubernetes.operator.savepoint.history.max.count`.

You are right, `disposeOnDelete` is false by default, and it can be controlled by the config `kubernetes.operator.savepoint.dispose-on-delete`.

I am not sure of the exact approach I will take to handle cleanup, but it will be implemented under [FLINK-35493](https://issues.apache.org/jira/browse/FLINK-35493) in another PR. That's a feature I felt we could easily separate to another PR, as this one is already growing quite big :D 

With a low periodic interval, there won't be any more load on the API server and etcd than tracking snapshots in the Flink resource CR (as the operator does currently). Periodically cleaning them up might affect the load on API server/etcd depending on what solution we will go with. But that will be done under FLINK-35493.

Reply:
I have done some investigation, and I think I will be able to easily implement this in the FlinkDeployment and FlinkSessionJob controller observe phase, just like the operator currently cleans up savepoints. I could just add a new `InformerEventSource` so we would be able to list `FlinkStateSnapshot` resources as secondary resources when we reconcile FlinkDeployment/FlinkSessionJob resources without querying the Kubernetes API server.

Once this PR is merged, I will create a new PR for this functionality.